pkg/handlers/generic/lifecycle/ccm/nutanix/handler.go

```diff
@@ -134,77 +117,56 @@ func (p *provider) Apply(
 		return fmt.Errorf("failed to get values for nutanix-ccm-config: %w", err)
 	}
 
-	// The configMap will contain the Helm values, but templated with fields that need to be filled in.
-	values, err = templateValues(clusterConfig, values)
-	if err != nil {
-		return fmt.Errorf("failed to template Helm values read from ConfigMap: %w", err)
-	}
-
-	hcp := &caaphv1.HelmChartProxy{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: caaphv1.GroupVersion.String(),
-			Kind:       "HelmChartProxy",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: cluster.Namespace,
-			Name:      "nutanix-ccm-" + cluster.Name,
-		},
-		Spec: caaphv1.HelmChartProxySpec{
-			RepoURL:   helmChart.Repository,
-			ChartName: helmChart.Name,
-			ClusterSelector: metav1.LabelSelector{
-				MatchLabels: map[string]string{clusterv1.ClusterNameLabel: cluster.Name},
-			},
-			ReleaseNamespace: defaultHelmReleaseNamespace,
-			ReleaseName:      defaultHelmReleaseName,
-			Version:          helmChart.Version,
-			ValuesTemplate:   values,
-		},
-	}
-	handlersutils.SetTLSConfigForHelmChartProxyIfNeeded(hcp)
-	if err = controllerutil.SetOwnerReference(cluster, hcp, p.client.Scheme()); err != nil {
-		return fmt.Errorf(
-			"failed to set owner reference on nutanix-ccm installation HelmChartProxy: %w",
-			err,
-		)
-	}
+	applier := addons.NewHelmAddonApplier(
+		addons.NewHelmAddonConfig(
+			p.config.defaultValuesTemplateConfigMapName,
+			defaultHelmReleaseNamespace,
+			defaultHelmReleaseName,
+		),
+		p.client,
+		helmChart,
+	).WithValueTemplater(templateValuesFunc(clusterConfig))
 
-	if err = client.ServerSideApply(ctx, p.client, hcp, client.ForceOwnership); err != nil {
+	if err = applier.Apply(ctx, cluster, p.config.DefaultsNamespace(), log); err != nil {
 		return fmt.Errorf("failed to apply nutanix-ccm installation HelmChartProxy: %w", err)
 	}
 
 	return nil
 }
 
-func templateValues(clusterConfig *apivariables.ClusterConfigSpec, text string) (string, error) {
-	helmValuesTemplate, err := template.New("").Parse(text)
-	if err != nil {
-		return "", fmt.Errorf("failed to parse Helm values template: %w", err)
-	}
+func templateValuesFunc(
+	clusterConfig *apivariables.ClusterConfigSpec,
```

Comment:
Have this take in the nutanix object rather than the whole cluseterConfig variable, it would make it clearer and a little neater.

Reply:
Good idea. Done.